### DISCUSSION TITLE
Anemeter TX174

### DISCUSSION THREAD

**Discussion by geraldFS** [arendst/Tasmota#10805] [Q&A] [answered]
Hello,

I would like to integrate an anemometer (TX174 from Technoline) with Tasmota, unfortunately I have no idea how to do it.

In another project (www.wettermonster.de) I found some code, would it be possible to use it?

greetings
Gerald

```
void readWeatherMeters() {

	windSpeed = (numRevsAnemometer / (interval * 60)) * 2.4;
	numRevsAnemometer = 0;

	Percipitation = 0.2794 * (numClicksRain * 60 / interval);
	numClicksRain = 0;

	int windDirectionVoltage = analogRead(A0);

	if (windDirectionVoltage >= 212 && windDirectionVoltage < 273)    {windDirection = "N";}
	else if (windDirectionVoltage >= 577 && windDirectionVoltage < 665) {windDirection = "NNE";}
	else if (windDirectionVoltage >= 483 && windDirectionVoltage < 577) {windDirection = "NE";}
	else if (windDirectionVoltage >= 929 && windDirectionVoltage < 943) {windDirection = "ENE";}
	else if (windDirectionVoltage >= 906 && windDirectionVoltage < 929) {windDirection = "E";}
	else if (windDirectionVoltage >= 943 && windDirectionVoltage < 1023){windDirection = "ESE";}
	else if (windDirectionVoltage >= 795 && windDirectionVoltage < 858) {windDirection = "SE";}
	else if (windDirectionVoltage >= 858 && windDirectionVoltage < 906) {windDirection = "SSE";}
	else if (windDirectionVoltage >= 665 && windDirectionVoltage < 748) {windDirection = "S";}
	else if (windDirectionVoltage >= 748 && windDirectionVoltage < 795) {windDirection = "SSW";}
	else if (windDirectionVoltage >= 348 && windDirectionVoltage < 399) {windDirection = "SW";}
	else if (windDirectionVoltage >= 399 && windDirectionVoltage < 483) {windDirection = "WSW";}
	else if (windDirectionVoltage >= 0 && windDirectionVoltage < 106)   {windDirection = "W";}
	else if (windDirectionVoltage >= 163 && windDirectionVoltage < 212) {windDirection = "WNW";}
	else if (windDirectionVoltage >= 106 && windDirectionVoltage < 163) {windDirection = "NW";}
	else if (windDirectionVoltage >= 273 && windDirectionVoltage < 348) {windDirection = "NNW";}
}

##PIN definition Windgeschwindigkeit
    pinMode(14, INPUT_PULLUP);
    attachInterrupt(14, countAnemometer, FALLING);
```

**Comment by barbudor** [comment, marked as answer]
Apparently the wind direction is provided by the analog input.
You can use Tasmota with the analog input of an ESP8266 (nodeMCU or mini-D1 like on the web site diagram). This will be published in the SENSOR MQTT message under the JSON field `"Analog": { "A0": value }`
Transformation into wind direction can then be performed either in your home automation system or on Tasmota using rules.

For windspeed it is not possible to give an definitive answer just by looking at the 2 lines of code but generally this would be connected to a GPIO declared as a "Counter" and you would then have to convert the number of pulse per time period into a wind speed

**Reply by geraldFS** [reply]
How should one create such a rule or how can one implement this?
Unfortunately I am a beginner and have not yet understood the powerful bulwark of rules and scripting.
I'm happy when I can connect my sensors and get the basic configuration.
But Tasmota is so powerful and offers such ingenious things that I don't even understand.

**Comment by geraldFS** [comment]
Yes, the anemometer simply has a reed contact built into it (1 peak per revolution).
The formula is recognizable:
windSpeed (km / h) = (numRevsAnemometer / (interval * 60)) * 2.4

**Reply by curzon01** [reply]
_Analog cup anemometer_ driver xsns_68.windmeter might be your friend
It's not build-in in any Tasmota version

You have to compile your own firmware by enable the sensor and following [customize your build](https://tasmota.github.io/docs/Compile-your-build/#customize-your-build) 

Use

```C
#define USE_WINDMETER
```
to compile Tasmota including this driver - read the docs how to do that.

Then use command [Sensor68](https://tasmota.github.io/docs/Commands/#sensor68) to setup your needs

**Reply by barbudor** [reply]
You should read the section about [rules](https://tasmota.github.io/docs/Rules/)
 Tasmota has 3 rulesets : Rule1, Rule2, Rule3
Each contains multiple rules in the form of `ON triggercondition DO something ENDON`
You can have multiple rules in a rule set such as `ON trigger1 DO something1 ENDON ON trigger2 DO something2 ENDON ...`

In your case I would a trigger on the teleperiod reporting for the ANALOG#A0 measure
Based on the value, you would set the appropriate value in a variable for subsequent MQTT publishing
As the trigger cannot include an AND condition nor a range of validity but only a <, <=, >, >= or != you can take advantage of the BREAK instruction that stops evaluating the ruleset once a rule with a met trigger condition has been found.
You so need to order the conditions:
The C code could have been rewritten:
```
     if (windDirectionVoltage < 106) {windDirection = "W";}
else if (windDirectionVoltage < 163) {windDirection = "NW";}
else if (windDirectionVoltage < 212) {windDirection = "WNW";}
else if (windDirectionVoltage < 273) {windDirection = "N";}
else if (windDirectionVoltage < 348) {windDirection = "NNW";}
else if (windDirectionVoltage < 399) {windDirection = "SW";}
else if (windDirectionVoltage < 483) {windDirection = "WSW";}
else if (windDirectionVoltage < 577) {windDirection = "NE";}
else if (windDirectionVoltage < 665) {windDirection = "NNE";}
else if (windDirectionVoltage < 748) {windDirection = "S";}
else if (windDirectionVoltage < 795) {windDirection = "SSW";}
else if (windDirectionVoltage < 858) {windDirection = "SE";}
else if (windDirectionVoltage < 906) {windDirection = "SSE";}
else if (windDirectionVoltage < 929) {windDirection = "E";}
else if (windDirectionVoltage < 943) {windDirection = "ENE";}
else {windDirection = "ESE";}
```
You would translate this into rules in a rule set. As the ruleset will be very long you can't enter it in a single command. So you can use the `Rule +ON ....` syntax to append a new rule into a rule set (only the 1st doesn't have +)
```
Rule1  ON tele-ANALOG#A0<106 DO var1 "W" BREAK
Rule1 +ON tele-ANALOG#A0<163 DO var1 "NW" BREAK
Rule1 +ON tele-ANALOG#A0<212 DO var1 "WNW" BREAK
....
Rule1 +ON tele-ANALOG#A0<943 DO var1 "ENE" BREAK
Rule1 +ON tele-ANALOG#A0 DO var1 "ESE" ENDON
```
Once all entered, `Rule1 on` will enable the whole ruleset
Now that the wind direction is in variable, you can publish it each time the variable is updated:
```
Rule2 ON var1#state DO publish stat/weatherstation/winddirection %value% ENDON
Rule2 on
```

This is only an exemple and other solutions could be used for using the value.
You could do similar for the wind speed for exemple and decide to publish them together on a single topic with a JSON
`publish stat/weatherstation/wind {"direction": %var1%,"speed": %var2%}`
 
But please go through the docs. There are a lot of information there and every word, every paragraph counts.

**Reply by geraldFS** [reply]
> _Analog cup anemometer_ driver xsns_68.windmeter might be your friend
> It's not build-in in any Tasmota version
> 
> You have to compile your own firmware by enable the sensor and following [customize your build](https://tasmota.github.io/docs/Compile-your-build/#customize-your-build)
> 
> Use
> 
> ```c
> #define USE_WINDMETER
> ```
> 
> to compile Tasmota including this driver - read the docs how to do that.
> 
> Then use command [Sensor68](https://tasmota.github.io/docs/Commands/#sensor68) to setup your needs

Thanks, with the Sensor68 the wind sensor works straight away.

Thank you!

**Reply by geraldFS** [reply]
> You should read the section about [rules](https://tasmota.github.io/docs/Rules/)
> Tasmota has 3 rulesets : Rule1, Rule2, Rule3
> Each contains multiple rules in the form of `ON triggercondition DO something ENDON`
> You can have multiple rules in a rule set such as `ON trigger1 DO something1 ENDON ON trigger2 DO something2 ENDON ...`
> 
> In your case I would a trigger on the teleperiod reporting for the ANALOG#A0 measure
> Based on the value, you would set the appropriate value in a variable for subsequent MQTT publishing
> As the trigger cannot include an AND condition nor a range of validity but only a <, <=, >, >= or != you can take advantage of the BREAK instruction that stops evaluating the ruleset once a rule with a met trigger condition has been found.
> You so need to order the conditions:
> The C code could have been rewritten:
> 
> ```
>      if (windDirectionVoltage < 106) {windDirection = "W";}
> else if (windDirectionVoltage < 163) {windDirection = "NW";}
> else if (windDirectionVoltage < 212) {windDirection = "WNW";}
> else if (windDirectionVoltage < 273) {windDirection = "N";}
> else if (windDirectionVoltage < 348) {windDirection = "NNW";}
> else if (windDirectionVoltage < 399) {windDirection = "SW";}
> else if (windDirectionVoltage < 483) {windDirection = "WSW";}
> else if (windDirectionVoltage < 577) {windDirection = "NE";}
> else if (windDirectionVoltage < 665) {windDirection = "NNE";}
> else if (windDirectionVoltage < 748) {windDirection = "S";}
> else if (windDirectionVoltage < 795) {windDirection = "SSW";}
> else if (windDirectionVoltage < 858) {windDirection = "SE";}
> else if (windDirectionVoltage < 906) {windDirection = "SSE";}
> else if (windDirectionVoltage < 929) {windDirection = "E";}
> else if (windDirectionVoltage < 943) {windDirection = "ENE";}
> else {windDirection = "ESE";}
> ```
> 
> You would translate this into rules in a rule set. As the ruleset will be very long you can't enter it in a single command. So you can use the `Rule +ON ....` syntax to append a new rule into a rule set (only the 1st doesn't have +)
> 
> ```
> Rule1  ON tele-ANALOG#A0<106 DO var1 "W" BREAK
> Rule1 +ON tele-ANALOG#A0<163 DO var1 "NW" BREAK
> Rule1 +ON tele-ANALOG#A0<212 DO var1 "WNW" BREAK
> ....
> Rule1 +ON tele-ANALOG#A0<943 DO var1 "ENE" BREAK
> Rule1 +ON tele-ANALOG#A0 DO var1 "ESE" ENDON
> ```
> 
> Once all entered, `Rule1 on` will enable the whole ruleset
> Now that the wind direction is in variable, you can publish it each time the variable is updated:
> 
> ```
> Rule2 ON var1#state DO publish stat/weatherstation/winddirection %value% ENDON
> Rule2 on
> ```
> 
> This is only an exemple and other solutions could be used for using the value.
> You could do similar for the wind speed for exemple and decide to publish them together on a single topic with a JSON
> `publish stat/weatherstation/wind {"direction": %var1%,"speed": %var2%}`
> 
> But please go through the docs. There are a lot of information there and every word, every paragraph counts.

Thank you, with the draft that you sent me, the wind direction indicator worked immediately, but it might turn out that this is integrated into a module ala Sensor68?

But the wind direction indicator is now running - Thank you very much!

**Reply by geraldFS** [reply]
Hello,

I have now adapted the code a little and the first step works, the wind direction is written in "var1".

Unfortunately, Rule2 that should send "var1" does not do it.
Nothing is received in Home Assistant.

Is there a possibility that the content of "var1" is written to a device called wind direction?
Maybe had two solutions for me,

a) the wind direction is shown on the Tasmota page
b) the values are transferred to HASS.


greetings
Gerald

Here is my Rule1 (monitoring and writing of var1)

Rule1 ON tele-ANALOG#A0<120 DO var1 O BREAK
Rule1 +ON tele-ANALOG#A0<220 DO var1 SO BREAK
Rule1 +ON tele-ANALOG#A0<330 DO var1 S BREAK
Rule1 +ON tele-ANALOG#A0<520 DO var1 NO BREAK
Rule1 +ON tele-ANALOG#A0<700 DO var1 SW BREAK
Rule1 +ON tele-ANALOG#A0<880 DO var1 N BREAK
Rule1 +ON tele-ANALOG#A0<950 DO var1 NW BREAK
Rule1 +ON tele-ANALOG#A0<1100 DO var1 W ENDON
Rule1 ON

**Reply by barbudor** [reply]
You can't have it written on the Web GUI unless you move from Rules to Scripting-Language (both are exclusive and need a custom compile). Only scripting has the ability to add stuff on the web page

You can publish to any MQTT topic using the Publish command.
You can add at the beginning of your rules : 
`ON var1#state DO Publish stat/%topic%/windDirection %value% ENDON`

So that each time var1 is updated, it will trigger a publish.
Topic is an example. I don't know about H.A. so I can't advise what topic would be the best.
